keystoremanager/keystoremanager.go

```diff
@@ -181,159 +208,202 @@ Note that since we only allow trust data to be downloaded over an HTTPS channel
+	}
+
+	// If there are certificates for this specific GUN, we already have trusted
+	// certificates for this GUN. Let's make sure to check that this root is
+	// signed by them
+	if len(certsForCN) != 0 {
+		logrus.Debugf("found %d valid certificates for %s", len(certsForCN), gun)
+		err = signed.VerifyRoot(root, 0, trustmanager.CertsToKeys(certsForCN))
+		if err != nil {
+			logrus.Debugf("failed to verify TUF data for: %s, %v", gun, err)
+			return &ErrValidationFail{Reason: "failed to validate integrity of roots"}
+		}
+	}
+
+	// Validate the integrity of the new root (does it have valid signatures)
+	err = signed.VerifyRoot(root, 0, trustmanager.CertsToKeys(allValidCerts))
+	if err != nil {
+		logrus.Debugf("failed to verify TUF data for: %s, %v", gun, err)
+		return &ErrValidationFail{Reason: "failed to validate integrity of roots"}
+	}
+
+	// Getting here means A) we had trusted certificates and both the
+	// old and new validated this root; or B) we had no trusted certificates but
+	// the new set of certificates has integrity (self-signed)
+	logrus.Debugf("entering root certificate rotation for: %s", gun)
+
+	var rotationFailure bool
+	// To support root certificate rotation, let trust only the certs present in root
+	// First, we delete all the old certificates that aren't present in the root
+	for certID, cert := range certsToRemove(certsForCN, allValidCerts) {
+		logrus.Debugf("removing certificate with certID: %s", certID)
+		err = km.trustedCertificateStore.RemoveCert(cert)
+		if err != nil {
+			logrus.Debugf("failed to remove trusted certificate with keyID: %s, %v", certID, err)
+			rotationFailure = true
+		}
+	}
+
+	// Now we add all the new certificates (even if they arleady already exist)
+	for _, cert := range allValidCerts {
+		err := km.trustedCertificateStore.AddCert(cert)
+		if err != nil {
+			// If the error is already exists we don't fail the rotation
+			if _, ok := err.(*trustmanager.ErrCertExists); ok {
+				continue
+			}
+			logrus.Debugf("error adding new trusted certificate for: %s, %v", gun, err)
+			rotationFailure = true
+		}
+		// Getting the CertID for debug only, don't care about the error
+		certID, _ := trustmanager.FingerprintCert(cert)
+		logrus.Debugf("adding trust certificate for with certID %s for %s", certID, gun)
+	}
+
+	if rotationFailure {
+		return &ErrRootRotationFail{Reason: "failed to rotate root keys"}
+	}
+
+	logrus.Debugf("Root validation succeeded for %s", gun)
+	return nil
+}
+
+func validRootLeafCerts(root *data.SignedRoot, gun string) ([]*x509.Certificate, error) {
+	// Get a list of all of the leaf certificates present in root
+	allLeafCerts, _ := parseAllCerts(root)
+	var validLeafCerts []*x509.Certificate
+
+	// Go through every leaf certificate and check that the CN matches the gun
+	for _, cert := range allLeafCerts {
+		// Validate that this leaf certificate has a CN that matches the exact gun
+		if cert.Subject.CommonName != gun {
```

Comment:
@aaronlehmann it can't be nil but can it be the empty object so CommonName would be `""`?